pro_tes/middleware/middleware_pipeline.py

```diff
@@ -0,0 +1,128 @@
+"""Set up the middleware pipeline."""
+import importlib
+import logging
+
+from flask import current_app
+
+logger = logging.getLogger(__name__)
+
+
+class MiddlewarePipeline:
+    """Middleware pipeline."""
+
+    def __init__(self, middlewares=None):
+        """Construct object instance."""
+        self.middlewares = middlewares or []
+        logger.info(f"middleware: {self.middlewares}")
+
+    def add_middleware(self, middleware):
+        """Add a middleware to the pipeline."""
+        self.middlewares.append(middleware)
+
+    def process_request(self, request, *args, **kwargs):
+        """Process the incoming request.
+
+        Args:
+            request: Incoming request object.
+            *args: Additional positional arguments.
+            ***kwargs: Additional keyword arguments.
+
+        Returns:
+            The modified request object.
+        """
+        for middleware in self.middlewares:
+            if isinstance(middleware, list):
+                inner_success = False
+                for m in middleware:
+                    try:
+                        request = m.set_request(
+                            request,
+                            *args,
+                            **kwargs
+                        )
+                        inner_success = True
+                        break
+                    except Exception as exc:  # pylint: disable=W0703
+                        logger.exception(
+                            "Error occurred in inner middleware: %s",
+                            exc
+                        )
+                if not inner_success:
+                    raise Exception("inner middleware failed")
+            else:
+                try:
+                    request = middleware.set_request(
+                        request,
+                        *args,
+                        **kwargs
+                    )
+                except Exception as exc:  # pylint: disable=W0703
+                    logger.exception(
+                        "Error occurred in middleware: %s",
+                        exc
+                    )
+        return request
+
+
+def load_middleware_instance(middleware_path):
+    """Load a middleware instance.
+
+    Args:
+        middleware_path: Middleware path in the form:
+            "module.submodule.MiddlewareClass".
+
+    Returns:
+        Middleware instance.
+    """
+    module_path, class_name = middleware_path.rsplit('.', 1)
+    module = importlib.import_module(module_path)
+    middleware_class = getattr(module, class_name)
+    return middleware_class()
+
+
+def load_middlewares_from_config(config: dict) -> list:
```

Comment:
I think this function as well as the function below might as well be methods of the class

Reply:
will fix it in the next commit.